### ISSUE TITLE
TRN-3 — Multi-dataloader balancing and metrics/logging (error tables, optional wandb)

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1577]
**Depends on:** TRN-1 (#1575), DATA-1 (#1569) · **Blocks:** TRN-5 (#1639), GATE-3 (#1585)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** Two bundled gaps in the training loop, both landing on TRN-1 (#1575)'s stage seams. **(1) Balancing.** Legacy multihead training has none: per-head datasets are flattened into one pool — `train_set = ConcatDataset([train_sets[head] for head in heads])` (`mace/cli/run_train.py:754`) — and a single `DataLoader` shuffles the concatenation (`:777-783`), so a small head is visited proportionally to its size and can be drowned by a large one. Per-head loaders exist but only for evaluation and plotting (`run_train.py:742-751`, reassembled as `train_<head>`/`valid_<head>` dicts at `:960-966`). Multiple DataLoaders with **balancing — up-sampling smaller datasets to the largest** ("multiplying up to the largest dataset") — must be supported, with Lightning's `CombinedLoader` as a *design reference only* (its `max_size_cycle`-style joint iteration): heavy training abstractions are explicitly rejected, so no Lightning dependency, ever. **(2) Metrics/logging.** Legacy lives in `mace/tools/train.py` (815 lines): the `MACELoss` torchmetric (`:614`) accumulates DDP-reducible states (`dist_reduce_fx="sum"/"cat"`) for energy/forces/**magforces**/stress/virials/dipole/polarizability deltas; `evaluate()` (`:581`) runs it per dataloader; per-head validation log lines (`valid_err_log`, `:49-155`) carry the head name in **all eleven** `log_errors` branches; the per-head wandb log dict is built at `:280-316`. End-of-run error tables live in `mace/tools/tables_utils.py` (262 lines; `create_error_table` at `:23` — PrettyTable, per-loader rows, optional wandb export, `skip_heads` at `:32,:36,:112-113`). wandb setup is `setup_wandb` (`mace/tools/scripts_utils.py:1095`), which logs only the keys named by `--wandb_log_hypers` out of `vars(args)` (and computes a full `args_dict_json` it never uses, `:1112`). This ticket ports the *semantics* of all of that onto the v1 explicit loop — reproduce the numbers and the log surface, never the structure.

**Interface & constraints:**

1. **Balancing semantics (up-sample to largest).** N per-head train loaders are iterated jointly; one epoch is defined by the largest dataset — smaller loaders cycle (restart with a fresh shuffle) until the largest is exhausted, so every head contributes the same number of optimizer-visible batches per epoch. Deterministic under the run seed: each restart reshuffles with a derived (seed, epoch, cycle) key, so runs are reproducible and cycles are not identical repeats. Implemented natively in `mace_torch` (design reference: Lightning `CombinedLoader` in `max_size_cycle` mode) — **no Lightning import**.
2. **The legacy-equivalent unbalanced mode remains available** (proportional sampling over the concatenation, matching `ConcatDataset` + single-shuffle semantics at `run_train.py:749,:772-778`, including `shuffle=True` and the seeded `torch.Generator` at `:778`). GATE-3 (#1585) compares `--engine v1` tiny-task losses/errors against legacy, and PAR-1 (#1572) re-runs the black-box multihead workflows — those comparisons need the legacy-equivalent mode. Mode selection is a field of CFG-1 (#1574)'s training config, not a hard-coded default flip.
3. **`drop_last` is regime-dependent and must be carried, not defaulted.** Legacy sets `drop_last=(not args.lbfgs)` on the per-head and concatenated train loaders (`run_train.py:741,:757,:777`): a full-batch L-BFGS run must see every configuration, a mini-batch run drops the ragged tail. The balanced iterator exposes this as a per-stage property so TRN-5 (#1639)'s LBFGS stage gets the whole dataset without a special case in the loader.
4. **Composition is loader/sampler-level, above the data backends.** DATA-1 (#1569)'s `DataBackend` Protocol is map-style with exact `__len__` and an `iter_range` sequential fast path; DDP sharding is a **sampler concern** (a backend never knows about ranks). The balanced iterator must compose per-loader samplers such that TRN-4 (#1578) can wrap each with a distributed sampler *without changing balancing semantics* — do not bake rank logic into the balancing itself.
5. **Metrics.** The v1 successor of `MACELoss` (`mace_torch/train/metrics.py`, seam created by TRN-1 (#1575)): distributed-reducible accumulation (sum counters + concatenated deltas, as legacy's `dist_reduce_fx` states at `train.py:618-651`), producing the same derived quantities as `MACELoss.compute()` (`:749-813`) — `loss`; MAE/RMSE/q95 of E (total and per-atom); F (MAE/RMSE/q95 plus `rel_mae_f`/`rel_rmse_f`); **magforces** (`mae_magf`, `rel_mae_magf`, `rmse_magf`, `rel_rmse_magf`, `q95_magf`); stress; virials (incl. per-atom RMSE); dipole (incl. per-atom and relative); polarizability (incl. per-atom). Evaluation runs per head over the valid-loader dict, exactly as legacy `train.py:196-213,:275-315`. **Known legacy wart, handle explicitly:** checkpoint selection uses only the *last* head's validation loss (`valid_loss = valid_loss_head  # consider only the last head for the checkpoint` — `train.py:214` and `:310`). v1 must make the head-aggregation rule for checkpoint selection an explicit, documented choice (config or constant) — neither silently reproduce nor silently change it; record the decision in the PR.
6. **Per-head logging.** Every validation log line carries the head name for every loss variant. This is already true in develop — all eleven `log_errors` branches of `valid_err_log` (`mace/tools/train.py:49-155`) format `head: {valid_loader_name}` — so it is a **test** gap, not a behaviour gap: pin it here so v1 cannot regress it. Eval results stay keyed per head (`train_<head>` / `valid_<head>` / per-test-set names) so multihead results associate each head with *its own* test set — the behavior implemented by `belongs_to_head` (`mace/cli/visualise_train.py:311`, pinned by `tests/unit/test_visualise_train.py`). TRN-3 (#1577) owns the per-head result data; the plot subcommand that renders it is CLI-1 (#1579).
7. **Error tables.** Port `create_error_table` semantics: one row per dataloader name, columns fixed by table type; the ten pinned types (legacy `--error_table` choices, `mace/tools/arg_parser.py:113-130`) are `PerAtomRMSE` (default), `TotalRMSE`, `PerAtomRMSEstressvirials`, `PerAtomMAEstressvirials`, `PerAtomMAE`, `TotalMAE`, `DipoleRMSE`, `DipoleMAE`, `DipolePolarRMSE`, `EnergyDipoleRMSE`. A skip-heads option succeeds `--skip_evaluate_heads` (comma list, `arg_parser.py:766`, split at `run_train.py:1233` — untested gap in legacy: add the test here). The P0-5 E2E contract pins that a table is printed at the end of training; column labels and units match legacy.
8. **wandb is optional** (the `wandb` extra): config fields succeeding the six legacy flags (`--wandb`, `--wandb_dir`, `--wandb_project`, `--wandb_entity`, `--wandb_name`, `--wandb_log_hypers`, `arg_parser.py:1124-1171`); import is guarded — wandb is never imported unless enabled (legacy imports it inside `setup_wandb`, `scripts_utils.py:1097`). When enabled: hyperparameters are logged from the **`ResolvedConfig`** (typed, from CFG-1 (#1574)), so the legacy `--wandb_log_hypers` allow-list over a raw namespace disappears and the whole resolved config is loggable; per-epoch per-head validation metrics are logged as legacy's `wandb_log_dict` (`train.py:280-316`); error tables export to wandb as `create_error_table(log_wandb=True)` does.
9. **Log setup.** The runtime-config successor of `--log_level` feeds the logger setup (legacy `tools.setup_logger(level, tag, directory, rank)` at `run_train.py:104`), rank-aware so TRN-4 (#1578)'s DDP port logs once per run, not once per rank.

**Task:**
1. `mace_torch/train/loaders.py`: the balanced multi-loader (up-sample to largest, per-cycle reshuffle keyed on (seed, epoch, cycle)) plus the legacy-equivalent proportional mode; per-stage `drop_last`; per-loader sampler seam left injectable for TRN-4 (#1578).
2. `mace_torch/train/metrics.py`: the `MACELoss` successor with distributed-reducible states and the full derived-quantity set including magforces; per-head `evaluate` over the valid-loader dict; head name in every validation log line.
3. Error tables: port the ten table types with legacy-identical columns/units, per-loader rows, skip-heads support; printed at end of training (P0-5 contract).
4. wandb integration behind the extra: guarded import, resolved-config hyperparameter logging, per-head epoch metrics, table export.
5. Wire balancing mode + logging into TRN-1 (#1575)'s `TrainStage` through the stage contract (config-driven; no argparse namespace).
6. Tests including the legacy gaps: per-head log lines for all eleven variants, skip-heads, balanced-count and determinism properties, metrics parity vs legacy `MACELoss` values on a shared fixture.

**Out of scope:** the DDP/distributed-sampler and SLURM ports and safetensors checkpointing (TRN-4 (#1578)); composable losses and per-stage schedules (TRN-2 (#1576)); the L-BFGS regime (TRN-5 (#1639)); the `mace plot training` subcommand and plot flags (CLI-1 (#1579)); dataset backends and statistics (DATA-1 (#1569)/2); replay-head data assembly (FT-1 (#1580)).

**Acceptance criteria:**
- [ ] With two heads of sizes n₁ < n₂, one balanced epoch yields the same number of batches from each head (small head up-sampled to the largest), verified by counting per-head samples; the unbalanced mode reproduces proportional `ConcatDataset` visitation.
- [ ] Same seed → identical batch sequence across runs; successive cycles of an up-sampled loader are reshuffled, not repeats. No Lightning import anywhere (asserted).
- [ ] A full-batch stage sees every configuration (`drop_last` off) while a mini-batch stage drops the ragged tail — the legacy `drop_last=(not lbfgs)` semantics, expressed per stage.
- [ ] v1 metrics match legacy `MACELoss`-derived values (loss, RMSE_E_per_atom, RMSE_F, relative F %, q95, magforces, stress/virials/dipole/polarizability variants) on a shared fixture at fp64.
- [ ] All ten error-table types render with legacy-identical columns and units; `PerAtomRMSE` is the default; skip-heads excludes the named heads; a table is printed at end of training (P0-5 workflow assertion passes under `--engine v1`).
- [ ] Every per-head validation log line contains the head name, for all eleven loss variants (test added — pins the develop behaviour that has no test today); multihead results keep each head associated with its own test set.
- [ ] wandb is not imported when disabled; when enabled, hyperparameters come from the resolved config and per-head epoch metrics are logged.
- [ ] The checkpoint-selection head-aggregation rule is explicit and documented (the last-head-only legacy wart is surfaced, not silently inherited).

**Inventory gaps assigned here:**
- [ ] `--skip_evaluate_heads` (`mace/tools/arg_parser.py:766`) — no test today.
- [ ] Per-head test parity plots and labels (#1445) and per-head validation logging (#1457): the behaviour is present in develop (`belongs_to_head`, the eleven `valid_err_log` branches) and `tests/unit/test_visualise_train.py` covers only `belongs_to_head` — the log-line side is still unpinned.
- [ ] `--wandb_log_hypers` — the allow-list that v1 removes; pin what legacy logs before removing it.

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/train/test_balanced_loaders.py -v   # counts, determinism, drop_last, no-lightning
python -m pytest packages/mace-torch/tests/train/test_metrics_tables.py -v     # MACELoss parity, 10 table types, per-head logs
python -m pytest tests/workflows -k "v1 and multihead" -n 2                    # E2E multihead under --engine v1
```
**Review focus:** the balancing iterator's determinism and its sampler seam (TRN-4 (#1578) must wrap it for DDP without changing which samples a head contributes), and metrics/table parity with legacy — these are the numbers GATE-3 (#1585) compares against `tiny_scaleshift_training_errors.json`, so a silent metric deviation invalidates the phase gate.

---
